File: rag/chunker.py
"""
Recursive character chunking.

Strategy: try to split on the strongest semantic boundary that keeps each piece
under `chunk_size`. Order of separators (strongest first):
    paragraph break -> line break -> sentence end -> clause -> word -> char

A sliding `chunk_overlap` is added between adjacent chunks so that a sentence
straddling a boundary still appears (in part) on both sides. This preserves
local context for retrieval and reduces "lost-at-the-seam" failures.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import List


DEFAULT_SEPARATORS: List[str] = ["\n\n", "\n", ". ", "? ", "! ", "; ", ", ", " ", ""]


@dataclass
class Chunk:
    text: str
    index: int
    char_start: int
    char_end: int


def _split_recursive(text: str, separators: List[str], chunk_size: int) -> List[str]:
    """Recursively split `text` so each piece is <= chunk_size when possible."""
    if len(text) <= chunk_size:
        return [text] if text else []

    sep = separators[0]
    rest = separators[1:]

    if sep == "":
        # Last resort: hard split by character.
        return [text[i : i + chunk_size] for i in range(0, len(text), chunk_size)]

    parts = text.split(sep)
    pieces: List[str] = []
    for part in parts:
        if not part:
            continue
        if len(part) <= chunk_size:
            pieces.append(part)
        else:
            pieces.extend(_split_recursive(part, rest, chunk_size))

    # Greedily merge adjacent pieces (re-attaching the separator) while staying
    # under chunk_size. This avoids ending up with hundreds of tiny fragments.
    merged: List[str] = []
    current = ""
    for piece in pieces:
        candidate = piece if not current else current + sep + piece
        if len(candidate) <= chunk_size:
            current = candidate
        else:
            if current:
                merged.append(current)
            current = piece
    if current:
        merged.append(current)
    return merged


def chunk_text(
    text: str,
    chunk_size: int = 1000,
    chunk_overlap: int = 150,
) -> List[Chunk]:
    """Split `text` into overlapping chunks suitable for embedding."""
    if chunk_overlap >= chunk_size:
        raise ValueError("chunk_overlap must be smaller than chunk_size")

    text = text.strip()
    if not text:
        return []

    raw = _split_recursive(text, DEFAULT_SEPARATORS, chunk_size)

    chunks: List[Chunk] = []
    cursor = 0  # tracks position in original text for char_start/char_end
    prev_tail = ""

    for i, piece in enumerate(raw):
        # Prepend tail of previous chunk for overlap context.
        body = (prev_tail + " " + piece).strip() if prev_tail else piece
        # Locate the piece's position in the original text from `cursor` onward
        # so char offsets remain accurate even when separators were stripped.
        found = text.find(piece, cursor)
        if found == -1:
            found = cursor
        char_start = max(0, found - len(prev_tail))
        char_end = found + len(piece)
        cursor = char_end

        chunks.append(
            Chunk(text=body, index=i, char_start=char_start, char_end=char_end)
        )

        # Build overlap tail for next iteration: last `chunk_overlap` chars of piece.
        prev_tail = piece[-chunk_overlap:] if chunk_overlap > 0 else ""

    return chunks

File: src/chunker.js
/**
 * Recursive character chunking.
 *
 * Strategy: split on the strongest semantic boundary that keeps each piece
 * under `chunkSize`. Order of separators (strongest first):
 *     paragraph break -> line break -> sentence end -> clause -> word -> char
 *
 * A sliding `chunkOverlap` is added between adjacent chunks so a sentence
 * straddling a boundary still appears (in part) on both sides. This preserves
 * local context for retrieval and reduces "lost-at-the-seam" failures.
 */

const DEFAULT_SEPARATORS = ['\n\n', '\n', '. ', '? ', '! ', '; ', ', ', ' ', ''];

/** Recursively split `text` so each piece is <= chunkSize when possible. */
function splitRecursive(text, separators, chunkSize) {
  if (text.length <= chunkSize) return text ? [text] : [];

  const sep = separators[0];
  const rest = separators.slice(1);

  if (sep === '') {
    // Last resort: hard split by character.
    const out = [];
    for (let i = 0; i < text.length; i += chunkSize) out.push(text.slice(i, i + chunkSize));
    return out;
  }

  const parts = text.split(sep);
  const pieces = [];
  for (const part of parts) {
    if (!part) continue;
    if (part.length <= chunkSize) pieces.push(part);
    else pieces.push(...splitRecursive(part, rest, chunkSize));
  }

  // Greedily merge adjacent pieces (re-attaching the separator) while staying
  // under chunkSize, so we don't end up with hundreds of tiny fragments.
  const merged = [];
  let current = '';
  for (const piece of pieces) {
    const candidate = current ? current + sep + piece : piece;
    if (candidate.length <= chunkSize) {
      current = candidate;
    } else {
      if (current) merged.push(current);
      current = piece;
    }
  }
  if (current) merged.push(current);
  return merged;
}

/**
 * Split `text` into overlapping chunks suitable for embedding.
 * Returns objects: { text, index, charStart, charEnd }.
 */
export function chunkText(text, { chunkSize = 1000, chunkOverlap = 150 } = {}) {
  if (chunkOverlap >= chunkSize) throw new Error('chunkOverlap must be smaller than chunkSize');

  text = text.trim();
  if (!text) return [];

  const raw = splitRecursive(text, DEFAULT_SEPARATORS, chunkSize);

  const chunks = [];
  let cursor = 0; // position in original text, for accurate char offsets
  let prevTail = '';

  raw.forEach((piece, i) => {
    const body = prevTail ? `${prevTail} ${piece}`.trim() : piece;
    // Locate the piece in the original text from `cursor` onward so offsets stay
    // accurate even though separators were stripped during splitting.
    let found = text.indexOf(piece, cursor);
    if (found === -1) found = cursor;
    const charStart = Math.max(0, found - prevTail.length);
    const charEnd = found + piece.length;
    cursor = charEnd;

    chunks.push({ text: body, index: i, charStart, charEnd });

    prevTail = chunkOverlap > 0 ? piece.slice(-chunkOverlap) : '';
  });

  return chunks;
}
